### DISCUSSION TITLE
vscode.workspace.fs.writeFile() seems to wipe the file when being called during shutdown, and how to avoid it

### DISCUSSION THREAD

**Discussion by scale-tone** [microsoft/vscode-discussions#548] [Extension Development QnA] [answered]
Hi everyone,

I'm doing [this extension](https://github.com/scale-tone/kee-shepherd), and what it does at VsCode's shutdown is that it tries to update some text files in the workspace, replacing whatever secrets can be there with what I call 'anchors'.

[The code that writes to a file is here](https://github.com/scale-tone/kee-shepherd/blob/main/kee-shepherd-vscode/src/KeeShepherdBase.ts#L56), and the [place where it gets invoked is here](https://github.com/scale-tone/kee-shepherd/blob/main/kee-shepherd-vscode/src/extension.ts#L223).

I know that during VsCode's shutdown all its APIs are unreliable (might throw), and that's somewhat OK for me - to make sure this 'stashing' operation eventually happens I'd just remember the fact that it didn't succeed and then reexecute it at next possible moment.

But what I've just noticed (on my Windows11 machine) is that occasionally some of those affected files end up being **empty**.
This happens very rarely, I myself have seen this only 2-3 times and so far unable to get a stable repro.

I have no explanation to this effect other than that it is caused by `vscode.workspace.fs.writeFile()` API being interrupted in the middle.

So my question is: can `vscode.workspace.fs.writeFile()` indeed be the culprit here? And if yes, what could be the workaround?

Thanks a lot for your help!

**Comment by isidorn** [comment, marked as answer]
Do you still see this issue with the latest version of VS Code? Have you found a solution? Thanks

**Reply by scale-tone** [reply]
Hi @isidorn , it was hardly noticeable anyway, and I didn't see any occurences of it since March.
Made [a test project](https://github.com/scale-tone/vscode-file-write-test/blob/main/src/extension.ts) to try magnifying the issue, played with it for a day - so far to no avail.

Do you think there could be some recent changes in VSCode, that might have fixed this?

Also, it could have been the problem not with `vscode.workspace.fs.writeFile()`, but with `vscode.workspace.fs.readFile()`. If during the shutdown it occasionally produces an empty (or somewhat inconsistent) array instead of throwing, then this effect can be explained. Do you think it looks more probable?

**Reply by isidorn** [reply]
There were changes in our FS layer that might have fixed it. We have also updated Electron that might have fixed it.
The truth is, I can not be certain. But I am glad to hear you are no longer seeing the issue.

**Reply by burkeholland** [reply]
Marking @isidorn dorn answer as answer. If this is not the case, please feel free to remove that.

**Reply by scale-tone** [reply]
Yes, that made a lot of sense, thanks, @isidorn !
